3ds2/src/main/java/com/adyen/checkout/adyen3ds2/DelegatedAuthentication.kt

```diff
@@ -0,0 +1,168 @@
+import android.content.Context
+import android.os.CountDownTimer
+import androidx.lifecycle.SavedStateHandle
+import com.adyen.authentication.AdyenAuthentication
+import com.adyen.authentication.AuthenticationLauncher
+import com.adyen.checkout.components.status.model.TimerData
+import com.adyen.checkout.core.log.LogUtil
+import com.adyen.checkout.core.log.Logger
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.filterNotNull
+import kotlinx.coroutines.launch
+import kotlin.time.Duration.Companion.milliseconds
+import kotlin.time.Duration.Companion.seconds
+
+@Suppress("TooManyFunctions")
+internal class DelegatedAuthentication(
+    private val savedStateHandle: SavedStateHandle
+) {
+
+    private var adyenAuthentication: AdyenAuthentication? = null
+
+    private var pendingRegistrationSdkInput: String?
+        get() = savedStateHandle[REGISTRATION_SDK_INPUT]
+        set(value) {
+            savedStateHandle[REGISTRATION_SDK_INPUT] = value
+        }
+    private var pendingAuthenticationSdkInput: String?
+        get() = savedStateHandle[AUTHENTICATION_SDK_INPUT]
+        set(value) {
+            savedStateHandle[AUTHENTICATION_SDK_INPUT] = value
+        }
+
+    private var countDownTimer: CountDownTimer? = null
+    internal var isTimerStarted: Boolean
+        get() = savedStateHandle[IS_TIMER_STARTED] ?: false
+        private set(value) {
+            savedStateHandle[IS_TIMER_STARTED] = value
+        }
+    private var currentMillisUntilFinished: Long
+        get() = savedStateHandle[MILLIS_UNTIL_FINISHED] ?: DEFAULT_TIMEOUT_IN_MILLIS
+        set(value) {
+            savedStateHandle[MILLIS_UNTIL_FINISHED] = value
+        }
+
+    private var _timeoutTimerFlow = MutableStateFlow<TimerData?>(null)
+    internal var timeoutTimerFlow = _timeoutTimerFlow.filterNotNull()
+
+    fun initialize(coroutineScope: CoroutineScope) {
+        if (isTimerStarted) {
+            startTimer(coroutineScope)
+        }
+    }
+
+    fun initAdyenAuthentication(context: Context, authenticationLauncher: AuthenticationLauncher) {
+        adyenAuthentication = AdyenAuthentication(context, authenticationLauncher)
+    }
+
+    fun getAdyenAuthentication(): AdyenAuthentication? = adyenAuthentication
+
+    internal suspend fun hasCredential(
+        application: Application,
+        sdkInput: String
+    ): Boolean? {
+        val hasDACredential = try {
+            AdyenAuthentication.hasCredential(application, sdkInput)
+        } catch (e: ClassNotFoundException) {
+            Logger.e(TAG, "hasCredential not executed because Authentication SDK is not present in project.")
+            null
+        } catch (e: NoClassDefFoundError) {
+            Logger.e(TAG, "hasCredential not executed because Authentication SDK is not present in project.")
+            null
+        } catch (e: IncompatibleClassChangeError) {
```

Comment:
As merchants need to present Authentication SDK dependency in addition to Drop-In/Components SDK, it might be thrown in case if the merchant include Authentication SDK version which is incompatible with Drop-In/Components SDK version

Reply:
Oh okay I see 👍